DQM/SiPixelHeterogeneous/plugins/SiPixelCompareRecHits.cc

```diff
@@ -0,0 +1,283 @@
+#include "DQMServices/Core/interface/MonitorElement.h"
+#include "DQMServices/Core/interface/DQMEDAnalyzer.h"
+#include "DQMServices/Core/interface/DQMStore.h"
+#include "DataFormats/Math/interface/approx_atan2.h"
+#include "DataFormats/SiPixelDetId/interface/PixelSubdetector.h"
+#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
+#include "DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsHost.h"
+#include "DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsSoA.h"
+#include "CUDADataFormats/TrackingRecHit/interface/TrackingRecHitSoAHost.h"
+#include "CUDADataFormats/TrackingRecHit/interface/TrackingRecHitsUtilities.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "Geometry/CommonDetUnit/interface/PixelGeomDetUnit.h"
+#include "Geometry/CommonTopologies/interface/PixelTopology.h"
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+
+template <typename T>
+class SiPixelCompareRecHits : public DQMEDAnalyzer {
```

Comment:
```suggestion
class SiPixelCompareRecHitsSoA : public DQMEDAnalyzer {
```

See [this comment](https://github.com/cms-sw/cmssw/pull/43964#issuecomment-1966013353).

Reply:
The issue here is that the CUDA-only comparison is already named like that. That is why I went with another option, although it wasn't a good one as it isn't explicit about the SoA format.

For tracks and vertices there is a work-around, as I named the classes/modules with the plural, but that is the only change. I think it would be very confusing to have `SiPixelPhase1CompareTrackSoA` and `SiPixelPhase1CompareTracksSoA`, even if just for some time, until the CUDA part is removed.

This is also valid for [suggestion below](https://github.com/cms-sw/cmssw/pull/43964#discussion_r1503823772)

Reply:
> The issue here is that the CUDA-only comparison is already named like that. That is why I went with another option, although it wasn't a good one as it isn't explicit about the SoA format.
> 
> For tracks and vertices there is a work-around, as I named the classes/modules with the plural, but that is the only change. I think it would be very confusing to have `SiPixelPhase1CompareTrackSoA` and `SiPixelPhase1CompareTracksSoA`, even if just for some time, until the CUDA part is removed.
> 
> This is also valid for [suggestion below](https://github.com/cms-sw/cmssw/pull/43964#discussion_r1503823772)

@fwyzard I just force pushed some changes, separated into the two commits you suggested. Do you have any comments/suggestions about this?

Reply:
@fwyzard could you please comment on my point above?

Reply:
> I think it would be very confusing to have `SiPixelPhase1CompareTrackSoA` and `SiPixelPhase1CompareTracksSoA`, even if just for some time, until the CUDA part is removed.

Yes, I agree that this would be very confusing.

Reply:
@fwyzard maybe I can add some "to do" comment to those files, stating that the names of the files and modules should be changed once the CUDA code is removed. Does this seems reasonable?

Reply:
Yes, that's a good idea.

Reply:
Done!